### DISCUSSION TITLE
Constrained optimisation with MMFF94

### DISCUSSION THREAD

**Discussion by KGrewal1** [rdkit/rdkit#5961] [Q&A] [answered]
Is it possible to perform a constrained optimisation, fixing specific bond lengths, angles, dihedrals (as opposed to fixing whole fragments)? As an example, would it be possible to constrain ethane to its eclipsed conformer before performing the optimisation. I've managed to scrape something together that seems to work, but can't tell if its using the forcefield class and allchem module correctly.
```
ethane = Chem.MolFromSmiles("CC")
mol = Chem.AddHs(ethane)
AllChem.EmbedMolecule(mol)
mp = AllChem.MMFFGetMoleculeProperties(mol, mmffVariant='MMFF94')
ff = AllChem.MMFFGetMoleculeForceField(mol, mp)
ff.MMFFAddTorsionConstraint(4,0,1,5,False,120,120,4e5)
ff.Minimize(maxIts=10000)
E0 = ff.CalcEnergy()

**Comment by ptosco** [comment, marked as answer]
@KGrewal1 That's correct. As a rule of thumb, it is generally good practice to set the torsion to the desired value with `rdMolTransforms.SetDihedralDeg()` before the minimization, rather than letting the minimizer do the job through the constraint. In fact, if the initial dihedral is far from the dihedral set in the constraint, the gradients will be huge and may lead to geometry distortions.
